File: port-spa-app/src/services/privacyPolicyService.ts
// Service for Privacy Policy API calls
import { apiClient } from './apiService';

export interface PrivacyPolicyDto {
    id: string;
    version: number;
    title: string;
    content: string;
    isCurrent: boolean;
    effectiveDate: string;
    createdAt: string;
    createdBy: string;
    changeReason?: string;
}

export interface UserPolicyStatusDto {
    requiresAcknowledgment: boolean;
    currentPolicy: PrivacyPolicyDto | null;
    lastAcknowledgedVersion?: number;
}

export interface AcknowledgePrivacyPolicyDto {
    policyId: string;
    policyVersion: number;
}

class PrivacyPolicyService {
    private readonly baseUrl = '/PrivacyPolicy';

    /**
     * Get the current active privacy policy (public endpoint - no auth required)
     */
    async getCurrentPolicy(): Promise<PrivacyPolicyDto | null> {
        try {
            const response = await apiClient.get<PrivacyPolicyDto>(`${this.baseUrl}/current`);
            return response.data;
        } catch (error: any) {
            if (error.response?.status === 404) {
                return null; // No policy published yet
            }
            throw error;
        }
    }

    /**
     * Get user's policy acknowledgment status (requires authentication)
     */
    async getUserPolicyStatus(): Promise<UserPolicyStatusDto> {
        const response = await apiClient.get<UserPolicyStatusDto>(`${this.baseUrl}/user/status`);
        return response.data;
    }

    /**
     * Acknowledge the current privacy policy (requires authentication)
     */
    async acknowledgePolicy(dto: AcknowledgePrivacyPolicyDto): Promise<void> {
        await apiClient.post(`${this.baseUrl}/acknowledge`, dto);
    }

    /**
     * Get all privacy policy versions (Admin only)
     */
    async getPolicyHistory(): Promise<PrivacyPolicyDto[]> {
        const response = await apiClient.get<PrivacyPolicyDto[]>(`${this.baseUrl}/history`);
        return response.data;
    }

    /**
     * Get specific policy by ID (Admin only)
     */
    async getPolicyById(id: string): Promise<PrivacyPolicyDto> {
        const response = await apiClient.get<PrivacyPolicyDto>(`${this.baseUrl}/${id}`);
        return response.data;
    }

    /**
     * Create new privacy policy (Admin only)
     */
    async createPolicy(title: string, content: string, changeReason?: string): Promise<PrivacyPolicyDto> {
        const response = await apiClient.post<PrivacyPolicyDto>(this.baseUrl, {
            title,
            content,
            changeReason
        });
        return response.data;
    }
}

export const privacyPolicyService = new PrivacyPolicyService();

File: PortProject.Api/Application/PrivacyPolicy/Services/PrivacyPolicyService.cs
using PortProject.Api.Application.PrivacyPolicy.DTOs;
using PortProject.Api.Domain.PrivacyPolicyAggregate;

namespace PortProject.Api.Application.PrivacyPolicy.Services
{
    public class PrivacyPolicyService : IPrivacyPolicyService
    {
        private readonly IPrivacyPolicyRepository _repository;

        public PrivacyPolicyService(IPrivacyPolicyRepository repository)
        {
            _repository = repository ?? throw new ArgumentNullException(nameof(repository));
        }

        public async Task<PrivacyPolicyDto> CreateAndPublishPolicyAsync(CreatePrivacyPolicyDto dto, string createdBy)
        {
            // Archive the current policy if exists
            var currentPolicy = await _repository.GetCurrentPolicyAsync();
            if (currentPolicy != null)
            {
                currentPolicy.MarkAsArchived();
                await _repository.UpdateAsync(currentPolicy);
            }

            // Create new policy
            var newPolicy = new Domain.PrivacyPolicyAggregate.PrivacyPolicy(
                dto.Title,
                dto.Content,
                createdBy,
                dto.ChangeReason
            );

            // Set version number
            var nextVersion = await _repository.GetNextVersionNumberAsync();
            newPolicy.SetVersion(nextVersion);
            newPolicy.MarkAsCurrent();

            // Save
            await _repository.AddAsync(newPolicy);

            return MapToDto(newPolicy);
        }

        public async Task<PrivacyPolicyDto> UpdateAndPublishPolicyAsync(Guid id, UpdatePrivacyPolicyDto dto)
        {
            var policy = await _repository.GetByIdAsync(id);
            if (policy == null)
                throw new KeyNotFoundException($"Privacy Policy with ID {id} not found.");

            policy.UpdateContent(dto.Title, dto.Content, dto.ChangeReason);
            await _repository.UpdateAsync(policy);

            return MapToDto(policy);
        }

        public async Task<PrivacyPolicyDto?> GetCurrentPolicyAsync()
        {
            var policy = await _repository.GetCurrentPolicyAsync();
            return policy == null ? null : MapToDto(policy);
        }

        public async Task<PrivacyPolicyDto?> GetPolicyByIdAsync(Guid id)
        {
            var policy = await _repository.GetByIdAsync(id);
            return policy == null ? null : MapToDto(policy);
        }

        public async Task<List<PrivacyPolicyDto>> GetPolicyHistoryAsync()
        {
            var policies = await _repository.GetPolicyHistoryAsync();
            return policies.Select(MapToDto).ToList();
        }

        public async Task<UserPolicyStatusDto> GetUserPolicyStatusAsync(string userId)
        {
            var currentPolicy = await _repository.GetCurrentPolicyAsync();
            var latestAcknowledgment = await _repository.GetUserLatestAcknowledgmentAsync(userId);

            var requiresAcknowledgment = false;
            if (currentPolicy != null)
            {
                if (latestAcknowledgment == null)
                {
                    requiresAcknowledgment = true;
                }
                else if (latestAcknowledgment.PolicyVersion < currentPolicy.Version)
                {
                    requiresAcknowledgment = true;
                }
            }

            return new UserPolicyStatusDto
            {
                RequiresAcknowledgment = requiresAcknowledgment,
                CurrentPolicy = currentPolicy == null ? null : MapToDto(currentPolicy),
                LastAcknowledgedVersion = latestAcknowledgment?.PolicyVersion
            };
        }

        public async Task AcknowledgePolicyAsync(string userId, AcknowledgePrivacyPolicyDto dto, 
            string? ipAddress = null, string? userAgent = null)
        {
            var policy = await _repository.GetByIdAsync(dto.PolicyId);
            if (policy == null)
                throw new KeyNotFoundException($"Privacy Policy with ID {dto.PolicyId} not found.");

            if (policy.Version != dto.PolicyVersion)
                throw new InvalidOperationException("Policy version mismatch.");

            var acknowledgment = new UserPolicyAcknowledgment(
                userId,
                dto.PolicyId,
                dto.PolicyVersion,
                ipAddress,
                userAgent
            );

            await _repository.AddAcknowledgmentAsync(acknowledgment);
        }

        private static PrivacyPolicyDto MapToDto(Domain.PrivacyPolicyAggregate.PrivacyPolicy policy)
        {
            return new PrivacyPolicyDto
            {
                Id = policy.Id,
                Version = policy.Version,
                Title = policy.Title,
                Content = policy.Content,
                IsCurrent = policy.IsCurrent,
                EffectiveDate = policy.EffectiveDate,
                CreatedAt = policy.CreatedAt,
                CreatedBy = policy.CreatedBy,
                ChangeReason = policy.ChangeReason
            };
        }
    }
}

